service/matching/matcher_test.go

```diff
@@ -210,55 +210,61 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
 	t.Equal(mustParent(t.queue.partition.(*tqid.NormalPartition), 20).RpcName(), req.GetTaskQueue().GetName())
 }
 
+//nolint:errcheck
 func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
-	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	intruptC := make(chan struct{})
 
-	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
-	happened, err := t.childMatcher.Offer(ctx, historyTask)
-	t.True(happened)
-	t.Nil(err)
+	// task waits for a local poller
+	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
 
-	intruptC := make(chan struct{})
-	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
-	time.Sleep(time.Millisecond)
+	var wg sync.WaitGroup
+	go func() {
+		t.rootMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:revive
+	}()
 
-	// should allow sync match when there is only young tasks in the backlog
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
-	t.Nil(err)
-	t.False(happened) // sync match did not happen, but we called the forwarder client
+	// Wait for the task to be added to the map
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
```

Comment:
why a busy loop when we can have Eventually? also that way you wont need the wg business.

Reply:
This was mostly for keeping things consistent with the way we were having sync mechanisms in the overall file (we use waitGroups and channels). I don't mind using Eventually.

Reply:
I think we do want to fail the test if this step takes many seconds, that would be very weird.
We usually use Eventually in the tests to wait for something async to happen, so I think changing to eventually would be more consistent to the way we write tests in general.
Lastly, we should avoid having busy loops in tests (and in code) as it will use a lot of CPU and might cause delays and flakiness to other tests that run in parallel.